File: compiler/src/Graph.swift
import OrderedCollections
import Synchronization

struct Graph {
    struct Edge: Hashable {
        var from: Node
        var to: Node
        var label: String
    }

    struct Group {
        var nodes: [Node]
        var types: [ConstructedType]
    }

    fileprivate var replacements: OrderedDictionary<Node, Node> = [:]
    fileprivate var edges: [Edge] = []
    var groups: [Group] = []

    mutating func replace(_ node: Node, with replacement: Node) {
        self.replacements[node] = replacement
    }

    mutating func edge(from: Node, to: Node, label: String) {
        self.edges.append(.init(from: from, to: to, label: label))
    }

    func serialize(including mask: OrderedSet<Node>, db: DB) -> SerializedGraph {
        // Ensure IDs are stable
        func id(for node: Node) -> String? {
            guard let span = db[node, Syntax.self]?.value.span else { return nil }
            return "node_\(span.description.replacing(/[^A-Za-z0-9]/, with: "_"))"
        }

        let maxIterations = 10

        func canDisplay(_ node: Node) -> Bool {
            (!node.isHidden || db.contains(Instantiated.self, for: node))
                && !db.contains(Defined.self, for: node)
        }

        func replacement(for node: Node) -> Node {
            var node = node
            while let replacement = self.replacements[node] { node = replacement }
            return node
        }

        var reachableNodes = OrderedDictionary<Node, Int?>(
            uniqueKeysWithValues: mask.map { ($0, nil) }
        )

        var inGroup: OrderedSet<String> = []
        var finishedGroups: OrderedSet<Int> = []
        var groups: OrderedDictionary<Int, (nodes: OrderedSet<Node>, types: [ConstructedType])> =
            [:]
        var edges: OrderedSet<Edge> = []
        for _ in 0..<maxIterations {
            var progress = false

            for (index, group) in self.groups.enumerated() {
                guard !finishedGroups.contains(index) else { continue }

                let nodes = OrderedSet(
                    group.nodes.lazy.filter(canDisplay(_:)).map(replacement(for:))
                        .filter { node in
                            guard let existing = reachableNodes[node] else { return false }
                            return existing == nil || existing == index
                        }
                )

                guard !nodes.isEmpty else { continue }

                if let (existingNodes, existingTypes) = groups[index] {
                    if existingNodes == nodes {
                        finishedGroups.append(index)
                    } else {
                        groups[index]!.nodes.formUnion(nodes)
                        progress = true
                    }

                    for type in existingTypes {
                        if !existingTypes.contains(type) { groups[index]!.types.append(type) }
                    }
                } else {
                    groups[index] = (nodes, group.types)
                    progress = true
                }

                inGroup.formUnion(nodes.compactMap(id(for:)))
                for node in nodes { reachableNodes[node] = index }
            }

            for edge in self.edges {
                let from = replacement(for: edge.from)
                let to = replacement(for: edge.to)

                guard reachableNodes[from] != nil || reachableNodes[to] != nil,
                    canDisplay(from) && canDisplay(to)
                else { continue }

                if reachableNodes[from] == nil { reachableNodes[from] = .some(nil) }
                if reachableNodes[to] == nil { reachableNodes[to] = .some(nil) }

                let edge = Edge(from: from, to: to, label: edge.label)
                if edges.append(edge).inserted { progress = true }
            }

            guard progress else { break }
        }

        // Remove groups containing only types

        func containsNonType(_ nodes: OrderedSet<Node>) -> Bool {
            nodes.contains { node in !db.contains(IsType.self, for: node) }
        }

        if groups.values.lazy.map(\.nodes).contains(where: containsNonType(_:)) {
            groups.removeAll { _, group in

                let keep = containsNonType(group.nodes)
                if !keep {
                    for node in group.nodes {
                        guard let id = id(for: node) else { continue }
                        inGroup.remove(id)
                    }
                }
                return !keep
            }
        }

        var result = SerializedGraph()

        var connectedNodes: OrderedSet<String> = []
        for edge in edges {
            guard let from = id(for: edge.from), let to = id(for: edge.to),
                inGroup.contains(from) && inGroup.contains(to)
            else { continue }

            connectedNodes.append(from)
            connectedNodes.append(to)

            result.edges.append(.init(from: from, to: to, label: edge.label))
        }

        // Remove unconnected nodes
        inGroup.removeAll { !connectedNodes.contains($0) }

        // Remove groups containing only function constants
        groups.removeAll { _, group in
            if group.nodes.count != 1 { return false }

            let node = group.nodes[0]
            guard let definitions = db[node, Resolved.self]?.definitions, definitions.count == 1,
                let definition = db[definitions[0], Defined.self]?.definition,
                definition is VariableDefinition, group.types.count == 1
            else { return false }

            if case .function = group.types[0].tag { return true }

            return false
        }

        for node in reachableNodes.keys {
            guard let id = id(for: node), inGroup.contains(id),
                let span = db[node, Syntax.self]?.value.span
            else { continue }

            result.nodes.append(.init(id: id, span: span))
        }

        for (var nodes, types) in groups.values {

            nodes.removeAll {
                guard let id = id(for: $0) else { return true }
                return !inGroup.contains(id)
            }

            guard !nodes.isEmpty else { continue }

            var labels = types.map { Type.constructed($0).display(db: db) }

            let conflict = labels.count != 1

            if labels.isEmpty { labels.append("_") }

            result.groups.append(
                .init(nodes: nodes.compactMap(id(for:)), labels: labels, conflict: conflict)
            )
        }

        return result
    }
}

public struct SerializedGraph {
    public struct Group {
        public var nodes: [String]
        public var labels: [String]
        public var conflict: Bool
    }

    public struct Node {
        public var id: String
        public var span: Span
    }

    public struct Edge {
        public var from: String
        public var to: String
        public var label: String
    }

    public var groups: [Group] = []
    public var nodes: [Node] = []
    public var edges: [Edge] = []

    public func toDOT() -> String {
        var graph = ""

        graph += "digraph G {\n"
        graph += "  node [shape=box, fontname=monospace]\n"

        for node in nodes { graph += "  \(node.id) [label=\(node.span.source.debugDescription)]\n" }

        for edge in edges {
            graph += "  \(edge.from) -> \(edge.to) [label=\(edge.label.debugDescription)]\n"
        }

        graph += "}"

        return graph
    }
}

File: compiler/core/src/graph.rs
use crate::{
    db::{Db, Node},
    facts::{GraphType, Syntax},
    span::Span,
    typecheck::{
        groups::update_type,
        instantiate::Instantiated,
        ty::{ConstructedTy, Ty, TyTag},
    },
    visit::definitions::{Defined, VariableDefinition},
};
use serde::{Deserialize, Serialize};
use std::{
    collections::{BTreeMap, BTreeSet},
    fmt::{self, Write},
};

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
struct GraphBuilderEdge {
    from: Node,
    to: Node,
    label: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GraphBuilderGroup {
    pub nodes: Vec<Node>,
    pub tys: Vec<ConstructedTy>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct GraphBuilder {
    replacements: BTreeMap<Node, Node>,
    edges: Vec<GraphBuilderEdge>,
    pub groups: Vec<GraphBuilderGroup>,
}

impl GraphBuilder {
    pub fn replace(&mut self, node: Node, replacement: Node) {
        self.replacements.insert(node, replacement);
    }

    pub fn edge(&mut self, from: Node, to: Node, label: impl Into<String>) {
        self.edges.push(GraphBuilderEdge {
            from,
            to,
            label: label.into(),
        });
    }

    pub fn group(&mut self, nodes: Vec<Node>, tys: Vec<ConstructedTy>) {
        self.groups.push(GraphBuilderGroup { nodes, tys });
    }

    pub fn build(&self, db: &Db, mask: &BTreeSet<Node>) -> Graph {
        const MAX_ITERATIONS: usize = 10;

        let mut reachable_nodes = mask
            .iter()
            .map(|&node| (node, None::<usize>))
            .collect::<BTreeMap<_, _>>();

        let mut in_group = BTreeSet::new();
        let mut finished_groups = BTreeSet::new();
        let mut groups = BTreeMap::<usize, (BTreeSet<Node>, Vec<ConstructedTy>)>::new();
        let mut edges = BTreeSet::<GraphBuilderEdge>::new();
        for _ in 0..MAX_ITERATIONS {
            let mut progress = false;

            for (index, group) in self.groups.iter().enumerate() {
                if finished_groups.contains(&index) {
                    continue;
                }

                let mut nodes = group
                    .nodes
                    .iter()
                    .copied()
                    .filter(|&node| self.can_display(db, node))
                    .map(|node| self.replacement_for(node))
                    .collect::<BTreeSet<_>>();

                nodes.retain(|node| {
                    reachable_nodes
                        .get(node)
                        .is_some_and(|existing| existing.is_none_or(|existing| existing == index))
                });

                if nodes.is_empty() {
                    continue;
                }

                let tys = group
                    .tys
                    .iter()
                    .map(
                        |ty| match update_type(db, &Ty::Constructed(ty.clone()), true) {
                            Ty::Constructed(ty) => ty,
                            _ => unreachable!(),
                        },
                    )
                    .collect::<Vec<_>>();

                if let Some((existing_nodes, existing_tys)) = groups.get_mut(&index) {
                    if *existing_nodes == nodes {
                        finished_groups.insert(index);
                    } else {
                        existing_nodes.extend(nodes.iter().cloned());
                        progress = true;
                    }

                    for ty in tys {
                        if !existing_tys.contains(&ty) {
                            existing_tys.push(ty);
                        }
                    }
                } else {
                    groups.insert(index, (nodes.clone(), tys));
                    progress = true;
                }

                in_group.extend(nodes.iter().flat_map(|&node| self.id(db, node)));
                reachable_nodes.extend(nodes.into_iter().map(|node| (node, Some(index))));
            }

            for edge in &self.edges {
                let from = self.replacement_for(edge.from);
                let to = self.replacement_for(edge.to);
                if !reachable_nodes.contains_key(&from) && !reachable_nodes.contains_key(&to) {
                    continue;
                }

                if !self.can_display(db, from) || !self.can_display(db, to) {
                    continue;
                }

                reachable_nodes.entry(from).or_insert(None);
                reachable_nodes.entry(to).or_insert(None);

                if edges.insert(GraphBuilderEdge {
                    from,
                    to,
                    label: edge.label.clone(),
                }) {
                    progress = true;
                }
            }

            if !progress {
                break;
            }
        }

        // Remove groups containing only types
        let contains_non_type =
            |nodes: &BTreeSet<Node>| nodes.iter().any(|&node| !db.contains::<GraphType>(node));
        if groups.values().any(|(nodes, _)| contains_non_type(nodes)) {
            groups.retain(|_, (nodes, _)| {
                let keep = contains_non_type(nodes);

                if !keep {
                    for node in nodes.iter().flat_map(|&node| self.id(db, node)) {
                        in_group.remove(&node);
                    }
                }

                keep
            });
        }

        let mut result = Graph::default();

        let mut connected_nodes = BTreeSet::new();
        for edge in edges {
            let Some(from) = self.id(db, edge.from) else {
                continue;
            };

            let Some(to) = self.id(db, edge.to) else {
                continue;
            };

            if !in_group.contains(&from) || !in_group.contains(&to) {
                continue;
            }

            connected_nodes.insert(from.clone());
            connected_nodes.insert(to.clone());

            result.edges.push(GraphEdge {
                from,
                to,
                label: edge.label,
            });
        }

        // Remove unconnected nodes
        in_group.retain(|id| connected_nodes.contains(id));

        for node in reachable_nodes.into_keys() {
            let Some(id) = self.id(db, node) else {
                continue;
            };

            if !in_group.contains(&id) {
                continue;
            }

            let Some(span) = db.get(node).map(|Syntax(syntax)| db.ast(syntax).span(db)) else {
                continue;
            };

            result.nodes.push(GraphNode {
                id,
                span: span.clone(),
            });
        }

        for (mut nodes, labels) in groups.into_values() {
            nodes.retain(|node| {
                let Some(id) = self.id(db, *node) else {
                    return false;
                };

                in_group.contains(&id)
            });

            if nodes.is_empty() {
                continue;
            }

            let mut labels = labels
                .into_iter()
                .map(|ty| GraphLabel {
                    kind: match ty.tag {
                        TyTag::Function | TyTag::Block => Some(String::from("function")),
                        _ => None,
                    },
                    display: Ty::Constructed(ty).display(db, true),
                })
                .collect::<Vec<_>>();

            let conflict = if labels.is_empty() {
                labels.push(GraphLabel {
                    kind: None,
                    display: String::from("_"),
                });
                true
            } else {
                labels.len() > 1
            };

            result.groups.push(GraphGroup {
                nodes: nodes
                    .into_iter()
                    .filter_map(|node| self.id(db, node))
                    .collect(),
                labels,
                conflict,
            })
        }

        result
    }

    fn id(&self, _db: &Db, node: Node) -> Option<String> {
        Some(format!(
            "node_{}",
            node.id()
                .to_string()
                .replace(|c: char| !c.is_alphanumeric(), "_")
        ))
    }

    fn can_display(&self, db: &Db, node: Node) -> bool {
        (!db.is_hidden(node) || db.contains::<Instantiated>(node))
            && db.get(node).is_none_or(|Defined(definition)| {
                // Allow variables, but not other definitions
                definition.downcast_ref::<VariableDefinition>().is_some()
            })
    }

    fn replacement_for(&self, mut node: Node) -> Node {
        while let Some(replacement) = self.replacements.get(&node) {
            node = *replacement;
        }

        node
    }
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct Graph {
    pub groups: Vec<GraphGroup>,
    pub nodes: Vec<GraphNode>,
    pub edges: Vec<GraphEdge>,
}

#[derive(Debug, Clone, Serialize)]
pub struct GraphGroup {
    pub nodes: Vec<String>,
    pub labels: Vec<GraphLabel>,
    pub conflict: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct GraphNode {
    pub id: String,
    pub span: Span,
}

#[derive(Debug, Clone, Serialize)]
pub struct GraphEdge {
    pub from: String,
    pub to: String,
    pub label: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct GraphLabel {
    pub kind: Option<String>,
    pub display: String,
}

impl Graph {
    pub fn write_dot(&self, mut w: impl Write) -> fmt::Result {
        writeln!(w, "digraph G {{")?;
        writeln!(w, "  node [shape=box, fontname=monospace]")?;

        for node in &self.nodes {
            writeln!(
                w,
                "  {} [label={}]",
                node.id,
                serde_json::to_string(node.span.source.as_str()).unwrap()
            )?;
        }

        for edge in &self.edges {
            writeln!(
                w,
                "  {} -> {} [label={}]",
                edge.from,
                edge.to,
                serde_json::to_string(&edge.label).unwrap()
            )?;
        }

        writeln!(w, "}}")?;

        Ok(())
    }
}
